compiler/src/dmd/common/smallbuffer.d

```diff
@@ -133,24 +133,25 @@ unittest
 `buffer` containing the converted string. The terminating zero is not part of the returned slice,
 but is guaranteed to follow it.
 */
-version(Windows) wchar[] toWStringz(const(char)[] narrow, ref SmallBuffer!wchar buffer) nothrow
+version(Windows) wchar[] toWStringz(scope const(char)[] narrow, ref SmallBuffer!wchar buffer) nothrow
 {
     import core.sys.windows.winnls : MultiByteToWideChar;
     import dmd.common.file : CodePage;
 
     if (narrow is null)
         return null;
 
-    const requiredLength = MultiByteToWideChar(CodePage, 0, narrow.ptr, cast(int) narrow.length, buffer.ptr, cast(int) buffer.length);
-    if (requiredLength < cast(int) buffer.length)
+    size_t length;
+    int i;
+    while (1)
     {
-        buffer[requiredLength] = 0;
-        return buffer[0 .. requiredLength];
+        // https://learn.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar
+        length = MultiByteToWideChar(CodePage, 0, narrow.ptr, cast(int) narrow.length, buffer.ptr, cast(int) buffer.length);
+        if (length < buffer.length)
+            break;
+        buffer.create(length + 1);
+        assert(++i == 1);   // ensure loop should only execute once or twice
```

Comment:
This code is arguably worse IMO. Instead of having the 2 calls clearly displayed, you have a loop for something that can only happen once or twice, and you have to add an assertion.

Using loops for things that may only happen once or twice is obfuscation, not simplification.

Reply:
I agree. Walter himself said in the past that "loops are bugs" and that straightforward code that reads from top to bottom is good.